sys/include/net/sock/addr.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2016 Freie Universität Berlin
+ *                    Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_sock_addr   Address abstractions
+ * @ingroup     net_sock
+ * @brief       Address abstractions for usage with @ref net_sock
+ * @{
+ *
+ * @file
+ * @brief   Address abstraction definitions for @ref net_sock
+ *
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ * @author  Kaspar Schleiser <kaspar@schleiser.de>
+ */
+#ifndef SOCK_ADDR_H_
+#define SOCK_ADDR_H_
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Special @ref net_netif "netif" ID for "any interface"
+ * @todo    Define in @ref net_netif
+ */
+#define SOCK_ADDR_ANY_NETIF (0)
+
+/**
+ * @brief   Address to bind to any IPv4 address
+ */
+#define SOCK_ADDR_IPV4_ADDR_ANY (0U)
+
+/**
+ * @brief   Address to bind to any IPv6 address
+ */
+#define SOCK_ADDR_IPV6_ADDR_ANY { { 0 } }
+
+typedef uint32_t sock_addr_ipv4_t;
+
+/**
+ * @brief   Type to represent an IPv6 address
+ */
+typedef struct {
+    uint8_t u8[16];         /**< bytes of the address */
+} sock_addr_ipv6_t;
```

Comment:
> You cut that out: "If that address gets casted to RIOT's ipv6_addr_t and then accessed using any of the non-uint8_t union members, that access is unaligned."

So what dos this have to with 6LoWPAN? This is just a common problem when working with unions. We could consider converting ipv6_addr_t to a simple struct, but this is a different discussion.

> Does it matter which type gets used in the API?

For me, yes, it does. One of the things that I dislike most about POSIX sockets and current `conn` are many additional data types.

> We tried to keep the interface simple 

IMO adding another (redundant) data type is exactly the contrary.

> and self-contained. 

I have the impression this is your main point, isn't it?

> RIOT's ipv6/addr.h is not exactly that.

IMO it is.

> That file alone has as many lines as all the dual-stack sock headers combined, and ~200 vs ~150 SLOC.

```
sloc sys/include/net/ipv6/addr.h
  Language  Files  Code  Comment  Blank  Total
     Total      1   198      468    131    786
```
198 SLOC.

Your dislike of documentation does not make it invalid.